crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2718,6 +2791,76 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
+    /// Performs lexicographic comparison between two slices of types.
+    ///
+    /// This comparison supports only equality and ordering operators: `==`, `!=`, `<`, `<=`, `>`, `>=`.
+    /// For lexicographic comparison, elements from both slices are compared pairwise using
+    /// `infer_binary_type_comparison`. If a conclusive result cannot be determined as a `BoolLiteral`,
+    /// it returns `bool`.
+    fn infer_lexicographic_type_comparison(
+        &mut self,
+        left: &[Type<'db>],
+        op: ast::CmpOp,
+        right: &[Type<'db>],
+    ) -> Option<Type<'db>> {
+        if !matches!(
+            op,
+            ast::CmpOp::Eq
+                | ast::CmpOp::NotEq
+                | ast::CmpOp::Lt
+                | ast::CmpOp::LtE
+                | ast::CmpOp::Gt
+                | ast::CmpOp::GtE
+        ) {
+            debug_assert!(
+                false,
+                "Unsupported comparison operator for lexicographic type comparison."
+            );
+            return None;
+        }
```

Comment:
I'd much prefer it if we could make the signature of this more type-safe; this makes me think that we should define a custom enum that this function accepts. Something like this? (The diff is relative to your PR branch)

```diff
--- a/crates/red_knot_python_semantic/src/types/infer.rs
+++ b/crates/red_knot_python_semantic/src/types/infer.rs
@@ -2709,15 +2709,16 @@ impl<'db> TypeInferenceBuilder<'db> {
                 let lhs_elements = lhs.elements(self.db).as_ref();
                 let rhs_elements = rhs.elements(self.db).as_ref();
 
+                let mut lexicographic_type_comparison =
+                    |op| self.infer_lexicographic_type_comparison(lhs_elements, op, rhs_elements);
+
                 match op {
-                    ast::CmpOp::Eq
-                    | ast::CmpOp::NotEq
-                    | ast::CmpOp::Lt
-                    | ast::CmpOp::LtE
-                    | ast::CmpOp::Gt
-                    | ast::CmpOp::GtE => {
-                        self.infer_lexicographic_type_comparison(lhs_elements, op, rhs_elements)
-                    }
+                    ast::CmpOp::Eq => lexicographic_type_comparison(RichCompareOperator::Eq),
+                    ast::CmpOp::NotEq => lexicographic_type_comparison(RichCompareOperator::Ne),
+                    ast::CmpOp::Lt => lexicographic_type_comparison(RichCompareOperator::Lt),
+                    ast::CmpOp::LtE => lexicographic_type_comparison(RichCompareOperator::Le),
+                    ast::CmpOp::Gt => lexicographic_type_comparison(RichCompareOperator::Gt),
+                    ast::CmpOp::GtE => lexicographic_type_comparison(RichCompareOperator::Ge),
                     ast::CmpOp::In | ast::CmpOp::NotIn => {
                         let mut eq_cnt = 0usize;
                         let mut not_eq_cnt = 0usize;
@@ -2757,7 +2758,7 @@ impl<'db> TypeInferenceBuilder<'db> {
                         // - `[ast::CmpOp::IsNot]`: returns `true` if the elements are definitely unequal, otherwise `bool`
                         let eq_result = self.infer_lexicographic_type_comparison(
                             lhs_elements,
-                            ast::CmpOp::Eq,
+                            RichCompareOperator::Eq,
                             rhs_elements,
                         );
                         // In Python, if `__eq__` is not defined, it implicitly uses object's `__eq__`. Adding `debug_assert!` as a precaution.
@@ -2800,25 +2801,9 @@ impl<'db> TypeInferenceBuilder<'db> {
     fn infer_lexicographic_type_comparison(
         &mut self,
         left: &[Type<'db>],
-        op: ast::CmpOp,
+        op: RichCompareOperator,
         right: &[Type<'db>],
     ) -> Option<Type<'db>> {
-        if !matches!(
-            op,
-            ast::CmpOp::Eq
-                | ast::CmpOp::NotEq
-                | ast::CmpOp::Lt
-                | ast::CmpOp::LtE
-                | ast::CmpOp::Gt
-                | ast::CmpOp::GtE
-        ) {
-            debug_assert!(
-                false,
-                "Unsupported comparison operator for lexicographic type comparison."
-            );
-            return None;
-        }
-
         // Compare paired elements from left and right slices
         for (l_ty, r_ty) in left.iter().zip(right.iter()) {
             let eq_result = self.infer_binary_type_comparison(*l_ty, ast::CmpOp::Eq, *r_ty);
@@ -2836,7 +2821,7 @@ impl<'db> TypeInferenceBuilder<'db> {
                     Truthiness::AlwaysTrue => continue,
                     // Types are not equal, perform the specified comparison and return the result
                     Truthiness::AlwaysFalse => {
-                        return self.infer_binary_type_comparison(*l_ty, op, *r_ty)
+                        return self.infer_binary_type_comparison(*l_ty, ast::CmpOp::Eq, *r_ty)
                     }
                     // If the intermediate result is abiguous, we cannot determine the final result as BoolLiteral.
                     // In this case, we simply return a bool instance.
@@ -2851,13 +2836,12 @@ impl<'db> TypeInferenceBuilder<'db> {
         let (l_len, r_len) = (left.len(), right.len());
 
         Some(Type::BooleanLiteral(match op {
-            ast::CmpOp::Eq => l_len == r_len,
-            ast::CmpOp::NotEq => l_len != r_len,
-            ast::CmpOp::Lt => l_len < r_len,
-            ast::CmpOp::LtE => l_len <= r_len,
-            ast::CmpOp::Gt => l_len > r_len,
-            ast::CmpOp::GtE => l_len >= r_len,
-            _ => return None,
+            RichCompareOperator::Eq => l_len == r_len,
+            RichCompareOperator::Ne => l_len != r_len,
+            RichCompareOperator::Lt => l_len < r_len,
+            RichCompareOperator::Le => l_len <= r_len,
+            RichCompareOperator::Gt => l_len > r_len,
+            RichCompareOperator::Ge => l_len >= r_len,
         }))
     }
 
@@ -3318,6 +3302,16 @@ impl<'db> TypeInferenceBuilder<'db> {
     }
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum RichCompareOperator {
+    Eq,
+    Ne,
+    Gt,
+    Ge,
+    Lt,
+    Le,
+}
+
```

I see this is similar to the idea you mentioned in https://github.com/astral-sh/ruff/pull/13712#discussion_r1798097932!

Reply:
And, we could avoid the `Option` in the return type, I think? Or, we could make it a hard panic instead of a `debug_assert` and add it to the function docs. Regardless of the solution, I'd prefer to avoid the return type to be an `Option` as I think the `None` variant is returned only for unsupported comparison operator.

Reply:
> I'd much prefer it if we could make the signature of this more type-safe; this makes me think that we should define a custom enum that this function accepts. Something like this? (The diff is relative to your PR branch)
> 
> ...
> 
> I see this is similar to the idea you mentioned in [#13712 (comment)](https://github.com/astral-sh/ruff/pull/13712#discussion_r1798097932)!


Looks great! I think this enum could also be useful for implementing inference for rich-comparisons. I’ve applied it!

Reply:
> And, we could avoid the `Option` in the return type, I think? Or, we could make it a hard panic instead of a `debug_assert` and add it to the function docs. Regardless of the solution, I'd prefer to avoid the return type to be an `Option` as I think the `None` variant is returned only for unsupported comparison operator.

Hmm, I don’t think we can avoid returning an `Option` in `infer_lexicographic_type_comparison`. Like `(1, "hello") < (1, 2)`, we need to propagate `None` from `infer_binary_type_comparison`. `None` can be returned in this line: [here](https://github.com/astral-sh/ruff/pull/13712/files#diff-65c2c229c88f4021638c996a7496384000d9e7b53b08426b34e92f120bd30b06R2807).

Reply:
> Hmm, I don’t think we can avoid returning an `Option` in `infer_lexicographic_type_comparison`. Like `(1, "hello") < (1, 2)`, we need to propagate `None` from `infer_binary_type_comparison`. `None` can be returned in this line: [here](https://github.com/astral-sh/ruff/pull/13712/files#diff-65c2c229c88f4021638c996a7496384000d9e7b53b08426b34e92f120bd30b06R2807).

Yes, this seems more tricky/requiring a larger refactor... I'm okay with leaving it for now, but maybe @dhruvmanila sees an easy solution :-)

Reply:
Returning `Option` here is correct. `None` signifies "unsupported comparison" and should emit a diagnostic. We may want to refactor to have `infer_binary_type_comparison`, and this method, return a `Result` instead of an `Option`, for the sake of a more informative diagnostic (that tells us precisely which two inner types can't be compared, rather than just telling us the two tuple types can't be compared), as I mentioned in another comment. But this wouldn't be a correctness fix, just a more-informative-diagnostics fix.